src/MongoDB.Driver.Core/Core/Bindings/ReadPreferenceBinding.cs

```diff
@@ -68,15 +69,18 @@ public ICoreSessionHandle Session
         public IChannelSourceHandle GetReadChannelSource(CancellationToken cancellationToken)
         {
             ThrowIfDisposed();
-            var server = _cluster.SelectServer(_serverSelector, cancellationToken);
+            var server = ServerSelectionHelper.SelectServer(_session, _cluster, _serverSelector, cancellationToken);
+            ServerSelectionHelper.PinShardRouterToSessionIfNeeded(_cluster, _session, server);
             return GetChannelSourceHelper(server);
         }
 
         /// <inheritdoc/>
         public async Task<IChannelSourceHandle> GetReadChannelSourceAsync(CancellationToken cancellationToken)
         {
             ThrowIfDisposed();
-            var server = await _cluster.SelectServerAsync(_serverSelector, cancellationToken).ConfigureAwait(false);
+            var server = await ServerSelectionHelper.SelectServerAsync(_session, _cluster, _serverSelector, cancellationToken)
+                .ConfigureAwait(false);
+            ServerSelectionHelper.PinShardRouterToSessionIfNeeded(_cluster, _session, server);
```

Comment:
Same as above.

And same for all below.

Reply:
Done, per previous discussion.